### DISCUSSION TITLE
How to use Sync-PnPSharePointUserProfilesFromAzureActiveDirectory with OnPremisesExtensionAttributes

### DISCUSSION THREAD

**Discussion by gabbsmo** [pnp/powershell#3131] [Q&A]
The documentation only mentions Graph extensions. Is there a way to use the built in on premises extensions attributes?

```
$user = Get-PnPAzureADUser -Identity "nobody@contoso.com" -Select OnPremisesExtensionAttributes
Sync-PnPSharePointUserProfilesFromAzureActiveDirectory -UserProfilePropertyMapping @{ "MyCustomProperty" = "ExtensionAttribute12" } -Folder "Import"
```

Error: Sync-PnPSharePointUserProfilesFromAzureActiveDirectory: Failed to create sync job. Ensure you're providing users to sync and that the mapping is correct.

Syncing with a standard property, such as DisplayName works fine.

**Comment by r998877** [comment]
I also want to know. Please can someone make a statement?

**Comment by AndersRask** [comment]
As answered elsewhere, the extension attributes are only added to your AAD schema when AD Connect is configured.

If you want these attributes populated, set up AD Connect.
There's a ton of posts online if you google how to set this up.
https://learn.microsoft.com/en-us/azure/active-directory/hybrid/connect/how-to-connect-sync-feature-directory-extensions

**Reply by r998877** [reply]
AAD Connect has already been set up and extension attributes are available in Azure AD.

However, Sync-PnPSharePointUserProfilesFromAzureActiveDirectory does not load those property values when used as shown in this example:
https://pnp.github.io/powershell/cmdlets/Sync-PnPSharePointUserProfilesFromAzureActiveDirectory.html#example-2

Note: ID in the attribute name (e.g. extension_b0b5aaa58a0a4287acd826c5b8330e48_CostCenter) has been replaced with the correct ID from my tenant. There is no error shown, the attribute is just not shown in the JSON file which is created.

As the documentation states, the user which should be synchronized should be retrieved from Azure AD first, e.g.

$users = Get-PnPAzureADUser -Identity "john.doe@contoso.com" -Select "DisplayName","EmployeeID", "MobilePhone", "OnPremisesExtensionAttributes"

So, the only way to get the extension attributes into the result is to -Select for "OnPremisesExtensionAttributes". Afterwards those attributes are accessible using 

$users[0].AdditionalProperties.OnPremisesExtensionAttributes.

But, how is this mapped to the mapping as explained in the documentation?

Sync-PnPSharePointUserProfilesFromAzureActiveDirectory -UserProfilePropertyMapping @{"CostCenter"="**extension_b0b5aaa58a0a4287acd826c5b8330e48_CostCenter**"} -Users $users

Any help is appreciated.

**Reply by AndersRask** [reply]
What do you see if you look into AAD extension attributes (either using AAD or PowerShell). Are the values set for the user?

Also check in Shared Documents for the site you have created a connection to, to verify the values in userprofilesyncdata-xxxx.json

**Reply by r998877** [reply]
I see filled extension attributes in the AAD portal. 

![image](https://github.com/pnp/powershell/assets/20636156/df33ad4b-2c8b-4891-8967-24e41504fa15)

When I try to get the value of extensionAttribute3 via PowerShell ...

`$users = Get-PnPAzureADUser -Identity $Config.TestUser -Select "DisplayName","EmployeeID", "MobilePhone", "extension_6cecf999ecb14aa7b3221e3d1c670160_extensionAttribute3"`

... it doesn't appear in the output (the ID in the attribute name is correct)

<img width="1103" alt="image" src="https://github.com/pnp/powershell/assets/20636156/36e1ad65-f8b7-4bf3-9340-d2caaeb70ec7">

[This post here](https://www.linkedin.com/pulse/updating-m365-user-profiles-delve-alex-g) does the same and the extensionAttributes are shown (see first and second screenshot in this post).

The only way I've figured out so far to get the extensionAttributes is to -Select for "OnPremisesExtensionAttributes"

<img width="884" alt="image" src="https://github.com/pnp/powershell/assets/20636156/aecdca1d-4112-4e0f-a6d3-4572ff26324c">

<img width="856" alt="image" src="https://github.com/pnp/powershell/assets/20636156/02dc4f11-832f-40aa-bbbe-a242f02ef2c9">

However, Sync-PnPSharePointUserProfilesFromAzureActiveDirectory does not take this into account.

When using the extension attribute in the sync mapping (as it should be)

`$SyncMapping = @{"CellPhone"="MobilePhone";"CostCenter"="extension_6cecf999ecb14aa7b3221e3d1c670160_extensionAttribute3"}`

and calling

`Sync-PnPSharePointUserProfilesFromAzureActiveDirectory -UserProfilePropertyMapping $SyncMapping -Users $users -Folder $Config.SPOJobDocLib`

The JSON file is created in the doc lib specified, but it does not contain a value for the extension attribute.

<img width="324" alt="image" src="https://github.com/pnp/powershell/assets/20636156/41e044eb-81bd-45c0-8328-1fed9400031a">

**Reply by r998877** [reply]
To add: I'm using PnP.PowerShell 2.2.0.

**Reply by AndersRask** [reply]
So the sync runs without errors? Then the mapping found the attribute, so it is weird that the json does not contain even an empty value.

Out of curiosity, what do you see if you use Get-AzureADUser from the AzureAD module (PoSh 5.1 only!)
$user = Get-AzureADUser -ObjectId $userPrincipalName
$user.ExtensionProperty

Do you see the extension attribute there, and does the field name (guid) match?

I'm no expert on the subject, but I don't think OnPremisesExtensionAttributes and directory extensions (ExtensionProperty) are the same thing.

I just set this up at a customer, and it worked fine with the properties that I verified in $user.ExtensionProperty.extension_11888abdaa4d47e392169cf1eb75ebd2_countryCode
but I don't have access to their tenant anymore, and my own dev tenant does not have extensions, so I can't test it here.

I found this article on the different extension attributes
https://merill.net/2023/02/azure-ad-extensions-and-attributes/

**Reply by r998877** [reply]
$user = Get-AzureADUser -ObjectId $userPrincipalName
$user.ExtensionProperty

results in 

<img width="800" alt="image" src="https://github.com/pnp/powershell/assets/20636156/04077816-9bc8-4259-a2a5-38b7724046e2">

So, it is empty.

Thanks for the link that explains different types of extensions. I'm referring to extensionAttributes 1-15, so I guess selecting for extension_6cecf999ecb14aa7b3221e3d1c670160_whatever only works for custom extensions, not for onPremisesExtensionAttributes)

As a workaround I will use 
https://pnp.github.io/powershell/cmdlets/New-PnPUPABulkImportJob.html
and create the json file beforehand on my own with with data from Get-PnPAzureADUser -Identity $Config.TestUser -Select "OnPremisesExtensionAttributes"

**Reply by AndersRask** [reply]
I havent dug enough into the PnP.Framework code that is used to fetch the user properties to know, if there is a way to get the onpremises extension attributes to work. Alternatively you could probably talk with the AD guys in your company about how the properties are mapped in AD Connect, and if it can be mapped to the custom extension properties.
